File: crates/pyqql/tests/test_dx.py
import unittest
import pyqql


class TestDxImprovements(unittest.TestCase):
    def test_prepared_statement_binding_and_compile_route(self):
        # 1. execute(Stmt, params=...) / Stmt.bind prepared statements
        stmt = pyqql.parse("QUERY :v FROM test_coll LIMIT :lim")[0]
        self.assertIn(":v", repr(stmt))
        self.assertEqual(str(stmt), "QUERY :v FROM test_coll LIMIT :lim")

        bound = stmt.bind({"v": [0.1, 0.2, 0.3], "lim": 5})
        self.assertEqual(
            str(bound), "QUERY VECTOR [0.1, 0.2, 0.3] FROM test_coll LIMIT 5"
        )

        route = stmt.compile_route(params={"v": [0.1, 0.2, 0.3], "lim": 5})
        self.assertEqual(route["method"], "POST")
        self.assertEqual(route["path"], "/collections/test_coll/points/query")
        self.assertEqual(route["payload"]["limit"], 5)
        self.assertEqual(len(route["payload"]["query"]["nearest"]), 3)

    def test_vector_truncation_for_readable_eyeball(self):
        # 7. bind() vector truncation for human readability
        vec = [0.1 * i for i in range(128)]
        s = pyqql.bind("QUERY :v FROM test_coll", {"v": vec}, truncate_vectors=True)
        self.assertIn("... (128 dims)", s)
        self.assertNotIn(str(vec[-1]), s)

    def test_dotted_and_nested_parameters(self):
        # 3. Dotted and nested parameter names
        nested_params = {"loc": {"lat": 12.34, "lon": 56.78}}
        s = pyqql.bind(
            "QUERY [0.1, 0.2] FROM test_coll WHERE lat = :loc.lat AND lon = :loc.lon",
            nested_params,
        )
        self.assertEqual(
            s,
            "QUERY [0.1, 0.2] FROM test_coll WHERE lat = 12.34 AND lon = 56.78",
        )

        flat_params = {"loc.lat": 12.34, "loc.lon": 56.78}
        s2 = pyqql.bind(
            "QUERY [0.1, 0.2] FROM test_coll WHERE lat = :loc.lat AND lon = :loc.lon",
            flat_params,
        )
        self.assertEqual(
            s2,
            "QUERY [0.1, 0.2] FROM test_coll WHERE lat = 12.34 AND lon = 56.78",
        )

    def test_execution_report_and_scored_point(self):
        # 4, 5, 8. ScoredPoint, ExecutionReport backward compatibility and typed accessors
        rep_dict = {
            "ok": True,
            "results": [
                {
                    "ok": True,
                    "operation": "QUERY",
                    "message": "Found 2 hits",
                    "data": [
                        {
                            "id": 936746218411023069,
                            "score": 0.95,
                            "payload": {"title": "Doc 1"},
                            "text": "Doc 1",
                            "collection": "coll_a",
                        },
                        {
                            "id": "c87bb3c1-a201-447a-8f5f-1555df27d14d",
                            "score": 0.82,
                            "payload": {"title": "Doc 2"},
                            "text": "Doc 2",
                            "collection": "coll_a",
                        },
                    ],
                },
                {
                    "ok": True,
                    "operation": "FACET",
                    "message": "Found 2 facet hit(s)",
                    "data": [
                        {"value": "tech", "count": 10},
                        {"value": "news", "count": 4},
                    ],
                },
                {
                    "ok": True,
                    "operation": "COUNT",
                    "message": "Count: 42",
                    "data": {"result": {"count": 42}},
                },
            ],
            "succeeded": 3,
            "failed": 0,
        }

        rep = pyqql.ExecutionReport(rep_dict)
        # Backward compatibility
        self.assertTrue(rep.ok)
        self.assertTrue(rep["ok"])
        self.assertEqual(rep.succeeded, 3)
        self.assertEqual(rep["succeeded"], 3)

        # 8. Typed ScoredPoint
        hits = rep.hits(0)
        self.assertEqual(len(hits), 2)
        # 5. ID integer vs string
        self.assertEqual(hits[0].id, 936746218411023069)
        self.assertIsInstance(hits[0].id, int)
        self.assertIsInstance(hits[1].id, str)
        self.assertEqual(hits[0].score, 0.95)
        self.assertEqual(hits[0]["title"], "Doc 1")
        self.assertEqual(hits[0].get("title"), "Doc 1")
        self.assertEqual(hits[0].collection, "coll_a")

        # 4. Facet normalized hits
        facet = rep.facet(1)
        self.assertEqual(len(facet), 2)
        self.assertEqual(facet[0]["value"], "tech")

        # Count accessor
        self.assertEqual(rep.count(2), 42)


if __name__ == "__main__":
    unittest.main()

File: crates/nqql-edge/test_dx.js
'use strict';

const assert = require('assert');
const nqql = require('./index.js');

console.log('Testing Node.js DX enhancements (nqql-edge)...');

// 1. Stmt.bind() and Stmt.toString()
{
  const stmts = nqql.parse('QUERY [1.0, 2.0] FROM docs WHERE status = :status LIMIT :limit');
  assert.strictEqual(stmts.length, 1);
  const stmt = stmts[0];

  assert.strictEqual(typeof stmt.bind, 'function');
  assert.strictEqual(typeof stmt.compileRoute, 'function');
  assert.strictEqual(typeof stmt.toString, 'function');

  // Test toString before binding
  const strBefore = stmt.toString();
  assert.ok(strBefore.includes(':status'));

  // Bind on statement
  const bound = stmt.bind({ status: 'active', limit: 10 });
  const strAfter = bound.toString();
  assert.ok(strAfter.includes("'active'"));
  assert.ok(strAfter.includes('LIMIT 10'));
  // Original stmt remains unchanged
  assert.ok(stmt.toString().includes(':status'));
}

// 2. Stmt.compileRoute(params)
{
  const stmts = nqql.parse('QUERY [0.1, 0.2] FROM items WHERE category = :cat LIMIT 5');
  const stmt = stmts[0];

  const route = stmt.compileRoute({ cat: 'books' });
  assert.strictEqual(route.method, 'POST');
  assert.ok(route.path.includes('/points/query'));
  assert.strictEqual(typeof route.payload, 'object');
  assert.strictEqual(route.payload.limit, 5);
}

// 3. Nested dictionary parameter expansion (:loc.lat, :loc.lon)
{
  const qql = 'QUERY [0.1, 0.2] FROM places WHERE lat = :loc.lat AND lon = :loc.lon';
  const bound = nqql.bind(qql, {
    loc: { lat: 37.7749, lon: -122.4194 }
  });
  assert.ok(bound.includes('37.7749'));
  assert.ok(bound.includes('-122.4194'));

  // On Stmt as well
  const stmt = nqql.parse(qql)[0];
  const boundStmt = stmt.bind({ loc: { lat: 37.7749, lon: -122.4194 } });
  assert.ok(boundStmt.toString().includes('37.7749'));
}

// 4. Vector truncation in bind() and Stmt.toString()
{
  const qql = 'QUERY [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] FROM docs WHERE id = :id';
  const boundTrunc = nqql.bind(qql, { id: 42 }, { truncateVectors: true });
  assert.ok(boundTrunc.includes('... (10 dims)'));

  const boundNoTrunc = nqql.bind(qql, { id: 42 }, { truncateVectors: false });
  assert.ok(!boundNoTrunc.includes('dims)'));
  assert.ok(boundNoTrunc.includes('[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]'));

  // Truncate vectors on Stmt.bind
  const stmt = nqql.parse(qql)[0];
  const boundStmt = stmt.bind({ id: 42 }, { truncateVectors: true });
  assert.ok(boundStmt.toString().includes('... (10 dims)'));
}

// 5. ExecutionReport and ScoredPoint
{
  const mockPayload = {
    ok: true,
    succeeded: 3,
    failed: 0,
    results: [
      {
        type: 'query',
        status: 'success',
        data: [
          { id: 1, score: 0.95, payload: { name: 'first' }, vector: [0.1, 0.2] },
          { id: 'uuid-2', score: 0.82, payload: { name: 'second' } }
        ]
      },
      {
        type: 'facet',
        status: 'success',
        data: [
          { value: 'red', count: 12 },
          { value: 'blue', count: 8 }
        ]
      },
      {
        type: 'count',
        status: 'success',
        message: 'Count: 42',
        data: { result: { count: 42 } }
      }
    ]
  };

  const report = new nqql.ExecutionReport(mockPayload);
  assert.strictEqual(report.ok, true);
  assert.strictEqual(report.succeeded, 3);
  assert.strictEqual(report.failed, 0);

  const hits = report.hits(0);
  assert.strictEqual(hits.length, 2);
  assert.strictEqual(hits[0].id, 1);
  assert.strictEqual(hits[0].score, 0.95);
  assert.strictEqual(hits[0].payload.name, 'first');
  assert.strictEqual(hits[0].get('name'), 'first');
  assert.deepStrictEqual(hits[0].vector, [0.1, 0.2]);
  assert.strictEqual(hits[1].id, 'uuid-2');
  assert.strictEqual(hits[1].score, 0.82);

  // Facet report
  assert.strictEqual(report.facet(1).length, 2);
  assert.strictEqual(report.facet(1)[0].value, 'red');

  // Count report
  assert.strictEqual(report.count(2), 42);
}

// 6. executeHits wrapper check
{
  assert.strictEqual(typeof nqql.executeHits, 'function');
  assert.strictEqual(typeof nqql.Client.prototype.executeHits, 'function');
}

console.log('All nqql-edge DX unit tests passed successfully!');
